src/sparsebase/utils/io/reader.cc

```diff
@@ -163,39 +397,120 @@ MTXReader<IDType, NNZType, ValueType>::ReadCOO() const {
 
           row[l] = m;
           col[l] = n;
-          vals[l] = w;
         }
 
         auto coo = new format::COO<IDType, NNZType, ValueType>(
-            M, N, L, row, col, vals, format::kOwned);
+            M, N, L, row, col, nullptr, format::kOwned);
         return coo;
-      } else {
-        // TODO: Add an exception class for this
-        throw ReaderException(
-            "Weight type for weighted graphs can not be void");
       }
-    } else {
+    } else if (symm == (int)MTXSymmetryOptions::symmetric || symm == (int)MTXSymmetryOptions::skew_symmetric) {
+      IDType *row = new IDType[L * 2];
+      IDType *col = new IDType[L * 2];
+      NNZType actual_nnzs = 0;
+      IDType m, n;
+      ValueType w;
+      if (weighted) {
+        if constexpr (!std::is_same_v<void, ValueType>) {
+          vals = new ValueType[L * 2];
+        }
+      }
       for (NNZType l = 0; l < L; l++) {
-        IDType m, n;
         fin >> m >> n;
+        if (weighted)
+          fin >> w;
 
         if (convert_to_zero_index_) {
           n--;
           m--;
         }
-
-        row[l] = m;
-        col[l] = n;
+        row[actual_nnzs] = m;
+        col[actual_nnzs] = n;
+        if (weighted)
+          vals[actual_nnzs] = w;
+        actual_nnzs++;
+        if (symm == (int)MTXSymmetryOptions::skew_symmetric || m != n) {
+          row[actual_nnzs] = n;
+          col[actual_nnzs] = m;
+          if (weighted)
+            vals[actual_nnzs] = w;
+          actual_nnzs++;
+        }
+      }
+      IDType *actual_rows = row;
+      IDType *actual_cols = col;
+      ValueType *actual_vals = vals;
+      if (symm == (int)MTXSymmetryOptions::symmetric && actual_nnzs != L * 2) {
+        actual_rows = new IDType[actual_nnzs];
+        actual_cols = new IDType[actual_nnzs];
+        memcpy(actual_rows, row, actual_nnzs * sizeof(IDType));
+        memcpy(actual_cols, col, actual_nnzs * sizeof(IDType));
+        delete[] row;
+        delete[] col;
+        if (weighted) {
+          actual_vals = new ValueType[actual_nnzs];
+          memcpy(actual_vals, vals, actual_nnzs * sizeof(ValueType));
+          delete[] vals;
+        }
       }
-
       auto coo = new format::COO<IDType, NNZType, ValueType>(
-          M, N, L, row, col, nullptr, format::kOwned);
+          M, N, actual_nnzs, actual_rows, actual_cols, actual_vals,
+          format::kOwned);
       return coo;
+    } else {
+      throw ReaderException("Reader only supports general, symmetric, and skew-symmetric symmetry options");
     }
   } else {
     throw ReaderException("file does not exists!!");
```

Comment:
We can also make a class out of this

Reply:
Actually I don't think these options are universal across readers, so I'm not sure it's worth making it its own class.